idaes/apps/grid_integration/bidder.py

```diff
@@ -0,0 +1,774 @@
+            implemented_power_output: realized power outputs: {unit: []}
+
+         Returns:
+             None
+        '''
+
+        for unit in m.UNITS:
+            m.pre_P_T[unit] = round(implemented_power_output[unit][-1],2)
+            m.pre_on_off[unit] = round(int(implemented_power_output[unit][-1] > 1e-3))
+
+        return
+
+    def update_model(self,implemented_power_output,implemented_shut_down, implemented_start_up):
+
+        '''
+        This method updates the parameters in the bidding model based on
+        the implemented power outputs, shut down and start up events.
+
+        Arguments:
+            implemented_power_output: realized power outputs: {unit: []}
+            implemented_shut_down: realized shut down events: {unit: []}.
+            implemented_start_up: realized start up events: {unit: []}
+
+         Returns:
+             None
+        '''
+
+        self._update_UT_DT(self.model,implemented_shut_down, implemented_start_up)
+        self._update_power(self.model,implemented_power_output)
+
+        return
+
+    # define a function to add bidding constraints to models
+    @staticmethod
+    def _add_bidding_constraints(m,NA_con_range = 24):
+
+        '''
+        This function takes the thermal model and add bidding cosntraints
+        to it in order to bid into DAM.
+
+        Arguments:
+            m: the bidding thermal generator model
+            NA_con_range: number of time steps this constraint will be enforced.
+
+        Returns:
+            None
+        '''
+
+        # generate scenarios combinations
+        scenario_comb = combinations(m.SCENARIOS,2)
+
+        # constraints for thermal generators
+        def bidding_con_fun1(m):
+            for k in scenario_comb:
+                for j in m.UNITS:
+                    for h in range(NA_con_range):
+                        yield (m.P_T[j,h,k[0]] - m.P_T[j,h,k[1]])\
+                        *(m.DAM_price[h,k[0]] - m.DAM_price[h,k[1]]) >= 0
+        m.bidding_con1 = pyo.ConstraintList(rule = bidding_con_fun1)
+
+        return
+
+    # define a function to add non_anticipativity constraints to models
+    @staticmethod
+    def add_non_anticipativity(m,NA_con_range = 2):
+
+        '''
+        This function takes the thermal model and add non non_anticipativity
+        cosntraints to it in order to do self-scheduling in DAM.
+
+        Arguments:
+            m: the bidding thermal generator model
+            NA_con_range: number of time steps this constraint will be enforced.
+
+        Returns:
+            None
+        '''
+
+        # generate scenarios combinations
+        scenario_comb = combinations(m.SCENARIOS,2)
+
+        def NAC_list_rules(m):
+            for k in scenario_comb:
+                for j in m.UNITS:
+                    for h in range(NA_con_range):
+                        yield m.gen_pow[j,h,k[0]] == m.gen_pow[j,h,k[1]]
+        m.NAC_SS_cons = ConstraintList(rule = NAC_list_rules)
+
+        return m
```

Comment:
This function returns `m`, but similar functions do not. Not sure if this is intentional.